thrall/app/lib/RetryHandler.scala

```diff
@@ -0,0 +1,59 @@
+package lib
+
+
+import akka.actor.ActorSystem
+import akka.pattern.{after, retry}
+import play.api.{Logger, MarkerContext}
+
+import scala.concurrent.{ExecutionContext, Future, TimeoutException}
+import scala.concurrent.duration.FiniteDuration
+import scala.util.{Failure, Success}
+
+object RetryHandler {
+  def handleWithRetryAndTimeout[T](f: () => Future[T],
+                                   retries: Int,
+                                   timeout: FiniteDuration,
+                                   delay: FiniteDuration
+                                  )(implicit actorSystem: ActorSystem,
+                                    executionContext: ExecutionContext,
+                                    mc: MarkerContext
+                                  ): () => Future[T] = {
+    def logFailures[T](f: () => Future[T])(
+      implicit executionContext: ExecutionContext, mc: MarkerContext
+    ): () => Future[T] = {
+      () => {
+        f().transform {
+          case Success(x) => Success(x)
+          case Failure(t: TimeoutException) => {
+            Logger.error("Failed with timeout. Will retry")
+            Failure(t)
+          }
+          case Failure(exception) => {
+            Logger.error("Failed with exception.", exception)
+            Failure(exception)
+          }
+        }
+      }
+    }
+
+    def handleWithTimeout[T](f: () => Future[T], attemptTimeout: FiniteDuration): () => Future[T] = () => {
+      val timeout = after(attemptTimeout, using = actorSystem.scheduler)(Future.failed(
+        new TimeoutException(s"Timeout of $attemptTimeout reached.")
+      ))
+      Future.firstCompletedOf(Seq(timeout, f()))
+    }
+
+    def handleWithRetry[T](f: () => Future[T], retries: Int, delay: FiniteDuration): () => Future[T] = () => {
+      implicit val scheduler = actorSystem.scheduler
+      var count = 0
+
+      def attempt = () => {
+        count = count + 1
+        Logger.info(s"Attempt $count of $retries")
```

Comment:
Would be good to add a bit more context to this log line to include the message subject, or the markers of the `UpdateMessage`. This way we can gain more insight into what gets retried for further investigation.

Reply:
The markercontext is passed in as an implicit, and these fields are present in kibana.